C/services/common/service_security.cpp

```diff
@@ -0,0 +1,316 @@
+					Logger::getLogger()->error("Name [%s]",  name.c_str());
+					if (name == sName)
+					{
+						serviceMatched = true;
+						break;
+					}
+				}
+
+				if (it->HasMember("type") &&
+				    (*it)["type"].IsString())
+				{
+					string type = (*it)["type"].GetString();
+					 Logger::getLogger()->error("Type [%s]",  type.c_str());
+					if (type == sType)
+					{
+						typeMatched = true;
+						break;
+					}
+				}
+			}
+		}
+		return serviceMatched == true || typeMatched == true;
+	}
+
+	return false;
+}
+
+/**
+ * Authentication Middleware for PUT methods
+ *
+ * Routine first check whether the service is configured with authentication
+ *
+ * Access bearer token is then verified against FogLAMP core API endpoint
+ * JWT token claims are passed to verifyURL and verifyService routines
+ *
+ * If access is granted the input funcPUT funcion is called
+ * otherwise error response is sent to the client
+ *
+ * @param response	The HTTP Response to send
+ * @param request	The HTTP Request
+ * @param funcPUT	The function to call in case of access granted
+ */
+void ServiceAuthHandler::AuthenticationMiddlewarePUT(shared_ptr<HttpServer::Response> response,
+				shared_ptr<HttpServer::Request> request,
+        			std::function<void(
+        		        shared_ptr<HttpServer::Response>,
+        		        shared_ptr<HttpServer::Request>)> funcPUT)
+{
+	bool acl_set = this->getAuthenticatedCaller();
+	if (acl_set)
+	{
+		Logger::getLogger()->debug("This service %s has AuthenticatedCaller flag set %d",
+			this->getName().c_str(),
+			acl_set);
+		map<string, string> tokenClaims;
+		bool ret = this->getMgmtClient()->verifyAccessBearerToken(request, tokenClaims);
+		if (!ret)
+		{
+			string responsePayload = QUOTE({ "error" : "invalid service bearer token"});
+			Logger::getLogger()->error("invalid service bearer token");
+			this->respond(response,
+					SimpleWeb::StatusCode::client_error_bad_request,
+					responsePayload);
+
+			return;
+		}
+
+		// TODO get token claims from FogLAMP issued bearer token
+		// Fill tokenClaims with examples
+		tokenClaims["aud"] = "Fledge";
+		tokenClaims["sub"] = "DISP1";		// Service Name
+		tokenClaims["iss"] = "Dispatcher";	// Service Type
+
+		// Check for valid service caller (name, type)
+		bool valid_service = this->verifyService(tokenClaims["sub"], tokenClaims["iss"]);
+		if (!valid_service)
+		{
+			string msg = "authorisation not granted to this service";
+			string responsePayload = "{ \"error\" : \"" + msg + "\" }";
+			Logger::getLogger()->error(msg);
+			this->respond(response,
+					SimpleWeb::StatusCode::client_error_unauthorized,
+					responsePayload);
+			return;
+		}
+
+		// Check URLS
+		bool access_granted = this->verifyURL(request->path, tokenClaims);
+		if (!access_granted)
+		{
+			string responsePayload = QUOTE({ "error" : "authorisation not granted to this resource"});
+			Logger::getLogger()->error("authorisation not granted to this resource");
```

Comment:
Should this be an error? We don't want to flood the logs or make things high level than they need to be